File: clients/android/app/src/main/java/app/dayflow/android/DayflowAIProviderConfiguration.kt
package app.dayflow.android

import android.net.Uri
import org.json.JSONObject
import java.net.URL

internal object DayflowAndroidEndpointPolicy {
    fun isAllowed(value: String): Boolean {
        val normalized = value.trim()
        if (normalized.isBlank()) return false
        val parsed = runCatching { URL(normalized) }.getOrNull() ?: return false
        val uri = Uri.parse(normalized)
        val scheme = parsed.protocol.lowercase()
        val host = uri.host?.lowercase() ?: return false
        val authority = uri.encodedAuthority ?: return false
        if (authority.contains('@') || uri.encodedQuery != null || uri.encodedFragment != null) return false
        return when (scheme) {
            "https" -> true
            "http" -> host == "localhost" || host == "127.0.0.1" || host == "::1"
            else -> false
        }
    }
}

object DayflowAIProviderIds {
    const val LOCAL = "local"
    const val GEMINI = "gemini"
    const val OPENAI_COMPATIBLE = "openai_compatible"
    const val CODEX = "chatgpt"
    const val CLAUDE = "claude"
}

data class DayflowAIProviderConfiguration(
    val providerId: String,
    val endpoint: String,
    val modelId: String,
) {
    val requiresApiKey: Boolean
        get() = providerId == DayflowAIProviderIds.GEMINI || providerId == DayflowAIProviderIds.OPENAI_COMPATIBLE

    val isConfigured: Boolean
        get() = providerId.isNotBlank() && modelId.isNotBlank() &&
            (providerId == DayflowAIProviderIds.CODEX || providerId == DayflowAIProviderIds.CLAUDE ||
                isSafeEndpoint(endpoint))

    private fun isSafeEndpoint(value: String): Boolean {
        return DayflowAndroidEndpointPolicy.isAllowed(value)
    }

    fun toJson(): String = JSONObject()
        .put("provider_id", providerId)
        .put("endpoint", endpoint)
        .put("model_id", modelId)
        .toString()

    companion object {
        fun fromJson(value: String): DayflowAIProviderConfiguration {
            val json = JSONObject(value)
            return DayflowAIProviderConfiguration(
                providerId = json.getString("provider_id"),
                endpoint = json.optString("endpoint", ""),
                modelId = json.getString("model_id"),
            )
        }
    }
}

/** DPAPI/Keystore equivalent for the on-device AI routing and API secret. */
class DayflowAIProviderStore(private val keyStore: DayflowAndroidKeyStore) {
    private val configurationName = "ai-provider-config-v1"
    private val apiKeyName = "ai-provider-api-key-v1"

    fun load(accountId: String): DayflowAIProviderConfiguration? =
        keyStore.loadText(accountId, configurationName)?.let { value ->
            runCatching { DayflowAIProviderConfiguration.fromJson(value) }.getOrNull()
        }

    fun loadApiKey(accountId: String): String? = keyStore.loadText(accountId, apiKeyName)

    fun save(accountId: String, configuration: DayflowAIProviderConfiguration, apiKey: String?) {
        require(configuration.isConfigured) { "The AI provider configuration is incomplete" }
        require(!configuration.requiresApiKey || !apiKey.isNullOrBlank()) { "This AI provider requires an API key" }
        keyStore.storeText(accountId, configurationName, configuration.toJson())
        if (apiKey.isNullOrBlank()) keyStore.delete(accountId, apiKeyName)
        else keyStore.storeText(accountId, apiKeyName, apiKey.trim())
    }

    fun clear(accountId: String) {
        keyStore.delete(accountId, configurationName)
        keyStore.delete(accountId, apiKeyName)
    }
}

File: clients/windows/Dayflow.Windows/Core/DayflowAIProviderConfiguration.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Dayflow.Windows.Core;

/// <summary>Provider identifiers shared by native clients without sharing API secrets.</summary>
public static class DayflowAIProviderIds
{
    public const string Local = "local";
    public const string Gemini = "gemini";
    public const string OpenAICompatible = "openai_compatible";
    public const string Codex = "chatgpt";
    public const string Claude = "claude";
}

internal static class DayflowWindowsEndpointPolicy
{
    public static bool IsAllowed(Uri endpoint)
    {
        if (string.IsNullOrWhiteSpace(endpoint.Host)
            || !string.IsNullOrEmpty(endpoint.UserInfo)
            || !string.IsNullOrEmpty(endpoint.Query)
            || !string.IsNullOrEmpty(endpoint.Fragment)) return false;

        if (string.Equals(endpoint.Scheme, Uri.UriSchemeHttps, StringComparison.OrdinalIgnoreCase)) return true;
        return string.Equals(endpoint.Scheme, Uri.UriSchemeHttp, StringComparison.OrdinalIgnoreCase)
            && endpoint.IsLoopback;
    }
}

public sealed record DayflowAIProviderConfiguration(
    [property: JsonPropertyName("provider_id")] string ProviderId,
    [property: JsonPropertyName("endpoint")] string Endpoint,
    [property: JsonPropertyName("model_id")] string ModelId)
{
    public bool RequiresApiKey => ProviderId is DayflowAIProviderIds.Gemini or DayflowAIProviderIds.OpenAICompatible;

    public bool IsConfigured => !string.IsNullOrWhiteSpace(ProviderId)
        && !string.IsNullOrWhiteSpace(ModelId)
        && (ProviderId is DayflowAIProviderIds.Codex or DayflowAIProviderIds.Claude
            || Uri.TryCreate(Endpoint, UriKind.Absolute, out var endpoint)
                && IsSafeEndpoint(endpoint));

    public static bool IsSafeEndpoint(Uri endpoint) => DayflowWindowsEndpointPolicy.IsAllowed(endpoint);
}

/// <summary>
/// Stores non-secret provider routing separately from the API secret. Both are
/// protected by the Windows DPAPI-backed local key store; neither is synced.
/// </summary>
public sealed class DayflowAIProviderStore
{
    private const string ConfigurationName = "ai-provider-config-v1";
    private const string ApiKeyName = "ai-provider-api-key-v1";
    private readonly DayflowWindowsKeyStore _keyStore;

    public DayflowAIProviderStore(DayflowWindowsKeyStore? keyStore = null)
    {
        _keyStore = keyStore ?? new DayflowWindowsKeyStore();
    }

    public DayflowAIProviderConfiguration? Load(string accountId)
    {
        var value = _keyStore.LoadText(accountId, ConfigurationName);
        if (string.IsNullOrWhiteSpace(value)) return null;
        try
        {
            return JsonSerializer.Deserialize<DayflowAIProviderConfiguration>(value);
        }
        catch (JsonException)
        {
            return null;
        }
    }

    public string? LoadApiKey(string accountId) => _keyStore.LoadText(accountId, ApiKeyName);

    public void Save(string accountId, DayflowAIProviderConfiguration configuration, string? apiKey)
    {
        if (!configuration.IsConfigured) throw new ArgumentException("The AI provider configuration is incomplete.", nameof(configuration));
        if (configuration.RequiresApiKey && string.IsNullOrWhiteSpace(apiKey)) throw new ArgumentException("This AI provider requires an API key.", nameof(apiKey));
        var json = JsonSerializer.Serialize(configuration);
        _keyStore.StoreText(accountId, ConfigurationName, json);
        if (string.IsNullOrWhiteSpace(apiKey))
        {
            _keyStore.Delete(accountId, ApiKeyName);
        }
        else
        {
            _keyStore.StoreText(accountId, ApiKeyName, apiKey.Trim());
        }
    }

    public void Clear(string accountId)
    {
        _keyStore.Delete(accountId, ConfigurationName);
        _keyStore.Delete(accountId, ApiKeyName);
    }
}
